include/pybind11/pytypes.h

```diff
@@ -1004,10 +1004,27 @@ class capsule : public object {
     PYBIND11_OBJECT_DEFAULT(capsule, object, PyCapsule_CheckExact)
     PYBIND11_DEPRECATED("Use reinterpret_borrow<capsule>() or reinterpret_steal<capsule>()")
     capsule(PyObject *ptr, bool is_borrowed) : object(is_borrowed ? object(ptr, borrowed) : object(ptr, stolen)) { }
-    explicit capsule(const void *value, void (*destruct)(PyObject *) = nullptr)
-        : object(PyCapsule_New(const_cast<void*>(value), nullptr, destruct), stolen) {
-        if (!m_ptr) pybind11_fail("Could not allocate capsule object!");
+
+    explicit capsule(const void *value)
+        : object(PyCapsule_New(const_cast<void *>(value), nullptr, nullptr), stolen) {
+        if (!m_ptr)
+            pybind11_fail("Could not allocate capsule object!");
+    }
+
+    explicit capsule(const void *value, void (*destructor)(void *)) {
```

Comment:
I was thinking it would be nice to make this take a function with a (deduced) `T *` argument, with a reinterpret cast of the capsule pointer to the deduced type when invoking the destructor.  Unfortunately I think that would require some special handling to accept a lambda, so might not be worthwhile.